### ISSUE TITLE
need a way to disable ggtags

### ISSUE THREAD

**Issue by emacs18** [syl20bnr/spacemacs#6819]
I typed "SPC h I" as instructed within my emacs, typed the message below, then hit C-c C-c.
However nothing seems to happen other than an *eww* buffer popping up with
a long URL starting with https://github.com/.

That is why I am creating this issue using the github web page.

#### Description :octocat:
On 2016/07/17, spacemacs/ggtags-mode-enable function was added on devel branch
to bunch of major mode hooks to enable ggtags in many modes. This change is
causing several problems.

One problem is that gtags simply is not robust enough to be used this way under
certain conditions such as on Red Hat 6. I often have to kill these processes
manually on Red Hat 6, because they eat up 100 percent of one CPU doing
who-knows-what for hours. I tried updating to newer GNU global few times, but
that did not help.

Another problem is that spacemacs fails to start if I remove certain ELPA
packages so that these are installed during emacs start up. Following is an
example stack trace where installing one of my own packages results in calling
"global" command which fails and shows the stack trace shown below.

#### Reproduction guide :beetle:
- Remove package-user-dir, i.e., all installed packages.
- Use default ~/.spacemacs except to add few packages to 
  dotspacemacs-additional-packages
- Start Emacs

*Observed behaviour:* :eyes: :broken_heart:
Stack trace shown above

*Expected behaviour:* :heart: :smile:
Emacs starts up without stack trace.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.1.1
- Spacemacs: 0.105.21
- Spacemacs branch: develop (rev. 36f3b63)
- Graphic display: t
- Distribution: spacemacs
- Editing style: hybrid
- Completion: helm
- Layers:
```elisp
((auto-completion :variables auto-completion-return-key-behavior 'complete auto-completion-tab-key-behavior 'cycle auto-completion-complete-with-key-sequence nil auto-completion-enable-help-tooltip t auto-completion-enable-sort-by-usage t auto-completion-private-snippets-directory nil)
 (c-c++ :variables c-c++-default-mode-for-headers 'c++-mode)
 clojure command-log emacs-lisp eyebrowse
 (git :variables git-magit-status-fullscreen t)
 github gtags html javascript latex markdown org perforce python shell shell-scripts helm spell-checking syntax-checking version-control my-navi)
```

#### Backtrace :paw_prints:
```
Debugger entered--Lisp error: (error "‘global’ non-zero exit: global: GTAGS not found.")
  signal(error ("‘global’ non-zero exit: global: GTAGS not found."))
  error("`%s' non-zero exit: %s" "global" "global: GTAGS not found.")
  ggtags-process-string("global" "-pr")
  ggtags-find-project()
  ggtags-after-save-function()
  run-hooks(after-save-hook)
  basic-save-buffer(nil)
  save-buffer()
  update-directory-autoloads("/u/kimr/opt/elpa25/kimr-doxymacs-20151026.1519")
  package-generate-autoloads(kimr-doxymacs "/u/kimr/opt/elpa25/kimr-doxymacs-20151026.1519")
  package--make-autoloads-and-stuff([cl-struct-package-desc kimr-doxymacs (20151026 1519) "No description available." nil tar "my-elpa" nil nil nil] "/u/kimr/opt/elpa25/kimr-doxymacs-20151026.1519")
  package-unpack([cl-struct-package-desc kimr-doxymacs (20151026 1519) "No description available." nil tar "my-elpa" nil nil nil])
  package--check-signature("/u/kimr/public_html/elpa/elpa/" "kimr-doxymacs-20151026.1519.tar" "kimr-doxymacs-20151026.1519 ...
  package-install-from-archive([cl-struct-package-desc kimr-doxymacs (20151026 1519) "No description available." nil tar "my-elpa" nil nil nil])
  mapc(package-install-from-archive ([cl-struct-package-desc kimr-doxymacs (20151026 1519) "No description available." nil tar "my-elpa" nil nil nil]))
  package-download-transaction(([cl-struct-package-desc kimr-doxymacs (20151026 1519) "No description available." nil tar "my-elpa" nil nil nil]))
  package-install(kimr-doxymacs)
  use-package-ensure-elpa(kimr-doxymacs)
  ...

```

**Comment by robbyoconnor** [comment]
Why not just exclude the package?

**Comment by emacs18** [comment]
That is not a solution.  Also enabling gtags so widely is undesirable just for the fact that it is so intrusive in adding hooks to to many major modes so that a command (i;.e., "global") is executed each time any file is saved even though I do not want GNU global in any of these files.

**Comment by TheBB** [comment]
Here's my suggestion.

* we don't enable `helm-gtags-mode` at all. Literally all this mode does is a keymap (which we supersede with our leader map) and an update in `after-save-hook`, which we don't need because `ggtags-mode` does it, anyway.
* we make a PRs to ggtags:
   * substitute user-error for error when the files can't be found, so we don't get backtraces every time
   * add option to disable automatic updating

Would that be sufficient?

**Comment by TheBB** [comment]
In fact ggtags claim to be quite optimized for large source trees. Can you disable helm-gtags-mode and see if ggtags alone works well? Assuming I am understanding correctly where your performance problems come from.

Maybe @leoliu can chime in? (see my previous comment in this thread)

**Comment by emacs18** [comment]
Not enabling gtags/ggtags by default I think is an ideal solution.

For me the issue is not helm-gtags-mode vs ggtags vs or some other GNU global related package.
The problem is that GNU global is not reliable at least on Red Hat 6.
Several times I get email from a colleague telling me that I'm monopolizing a sever with 16 cores.
When I run ps, I see dozens of processes running gtags all of which were apparently launched
via after-save-hook.  Many of these are eating up 100 percent of one core.  I then have to kill all
these gtags processes manually.  This is why I explicitly disable any automatic action by GNU global.

**Comment by TheBB** [comment]
It sounds like all that is problematic for you is the tags update, so providing an option to disable that would be sufficient for you.

**Comment by emacs18** [comment]
That would certainly be a step forward for me.  However I would rather not have gtags/ggtags enabled at all for 99 percent of all git projects since I don't need or use GNU global for them.  Hence not adding stuff to my major mode hooks, after-save-hook, etc., would be highly desirable.  At least having a way for me to opt out of spacemacs adding things that add complexity, add more CPU usage, more disk usage, etc., would be good.

**Comment by TheBB** [comment]
Fixed by #6829 

> At least having a way for me to opt out of spacemacs adding things that add complexity, add more CPU usage, more disk usage, etc., would be good. 

You have numerous opt-outs and opt-ins with layers and packages, I don't think we need more except in very specific circumstances. You must have opted in to gtags in the first place.

**Comment by emacs18** [comment]
Thanks for the fix.  I'll try it as soon as it is available.

Yes I opted in to enable ggtags layer.
What I expect from a layer is to setup autoloads mostly and so I can choose to use parts of
it as I see fit.  That was the case until July 17 when changes were made to enable ggtags
via many major mode hooks.

I want to use ggtags when I need it.  That is why I enabled ggtags layer.  The issue I'm reporting
the the July 17 change which made ggtags layer pretty much unusable for me.

**Comment by TheBB** [comment]
If the problem was that it was enabled for too many major modes, then you could already have removed them using the `:disabled-for` keyword in the layer list.

**Comment by TheBB** [comment]
The July 17 change doesn't really add any more complexity, it just expands gtags to use more major modes. If that's the kind of complexity you want to opt out from it has been in Spacemacs for a good while (search for `:disabled-for`).

**Comment by emacs18** [comment]
Thanks for sharing :disabled-for which I did not about before.
So you seem to be suggesting that I add something like the following
to enable gtags layer only for c-c++ layer:

```
 (gtags :disabled-for clojure emacs-lisp javascript latex python shell-scripts)
```

This will certainly help although I'm not sure whether this will solve
the start up problem I described in my original posting.

However this still does not give me the flexibility that I need.
Before July 17 change it was up to me to add (ggtags-mode 1) within my
own major mode hooks as I see fit.  I enable ggtags-mode not merely
by major mode, but also by other parameters such as project name or
type, i.e., I enable ggtags-mode only within *some* c++ projects and not
others.  This flexibility I was and is very useful to me.

That is why I would probably use the following code that is based off of
your documentation in PR #6829:

```
(setq-default dotspacemacs-configuration-layers
  '((gtags :variables gtags-enable-by-default nil)))
```

Today I do the following to achieve the same effect:

```
(advice-add 'spacemacs/ggtags-mode-enable :around #'ignore)
```

This is a hack that I would rather not use.
